File: app/src/main/java/net/itsjustsomedude/tokens/db/Event.java
package net.itsjustsomedude.tokens.db;

import androidx.room.Entity;
import androidx.room.PrimaryKey;

import java.util.Calendar;

@Entity
public class Event {
    public static final int DIRECTION_SENT = 1;
    public static final int DIRECTION_RECEIVED = 2;

    @PrimaryKey(autoGenerate = true)
    public Long id;

    public Calendar time;
    public int count;
    public String coop;
    public String kevId;
    public String person;
    public int direction;
    public int notification;

    public Event(String coop, String kevId, Calendar time, int count, String person, int direction, int note) {
        this.id = 0L;
        this.coop = coop;
        this.kevId = kevId;
        this.time = time;
        this.count = count;
        this.person = person;
        this.direction = direction;
        this.notification = note;
    }

    public Event(String coop, String kevId, Calendar time, int count, String person, int direction) {
        this(coop, kevId, time, count, person, direction, 0);
    }
}

File: app/src/main/java/net/itsjustsomedude/tokens/db/Event.kt
package net.itsjustsomedude.tokens.db

import androidx.room.Entity
import androidx.room.PrimaryKey
import java.util.Calendar
import kotlin.math.max
import kotlin.math.pow

@Entity
data class Event(
	@PrimaryKey(autoGenerate = true)
	val id: Long = 0,
	val coop: String,
	val kevId: String,

	val time: Calendar,
	val count: Int,
	val person: String,
	val direction: Int,
	val notification: Int = 0,

	val receiver: Int = 0
) {
	/**
	 * Start and End should be Unix values, not millis.
	 */
	fun tval(start: Long, end: Long): Double {
//        double duration = (end.getTimeInMillis() / 1000d) - (start.getTimeInMillis() / 1000d);
//        double elapsed = (time.getTimeInMillis() / 1000d) - (start.getTimeInMillis() / 1000d);

		val duration = (start - end).toDouble()
		val elapsed = (time.timeInMillis / 1000.0) - start

		val i = (1 - 0.9 * (elapsed / duration)).pow(4.0)
		val singleValue = max(i, 0.03)

		return singleValue * count
	}

	companion object {
		const val DIRECTION_SENT = 1
		const val DIRECTION_RECEIVED = 2
	}
}